include/aspect/particle/property/cpo.h

```diff
@@ -0,0 +1,361 @@
+           * derivatvies. Intended mostly for use in this plugin.
+           */
+          void
+          load_particle_data_extended(unsigned int cpo_index,
+                                      const ArrayView<double> &data,
+                                      std::vector<unsigned int> &deformation_type,
+                                      std::vector<double> &volume_fraction_mineral,
+                                      std::vector<std::vector<double>> &volume_fractions_mineral,
+                                      std::vector<std::vector<Tensor<2,3> > > &a_cosine_matrices_mineral,
+                                      std::vector<std::vector<double> > &volume_fractions_mineral_derivatives,
+                                      std::vector<std::vector<Tensor<2,3> > > &a_cosine_matrices_mineral_derivatives) const;
+
+          /**
+           * Stores information in variables into the data array. Intended for use by other parts of ASPECT.
+           */
+          static
+          void
+          store_particle_data(unsigned int cpo_data_position,
+                              const ArrayView<double> &data,
+                              std::vector<unsigned int> &deformation_type,
+                              std::vector<double> &volume_fraction_mineral,
+                              std::vector<std::vector<double>> &volume_fractions_mineral,
+                              std::vector<std::vector<Tensor<2,3> > > &a_cosine_matrices_mineral);
+
+
+          /**
+           * Stores information in variables into the data array. An extention of load_particle_data which includes the grain
+           * derivatvies. Intended mostly for use in this plugin.
+           */
+          void
+          store_particle_data_extended(unsigned int cpo_data_position,
+                                       const ArrayView<double> &data,
+                                       std::vector<unsigned int> &deformation_type,
+                                       std::vector<double> &volume_fraction_mineral,
+                                       std::vector<std::vector<double>> &volume_fractions_mineral,
+                                       std::vector<std::vector<Tensor<2,3> > > &a_cosine_matrices_mineral,
+                                       std::vector<std::vector<double> > &volume_fractions_mineral_derivatives,
+                                       std::vector<std::vector<Tensor<2,3> > > &a_cosine_matrices_mineral_derivatives) const;
+
+          /**
+           * Updates the volume fractions and rotation matrices with a Forward Euler scheme.
+           */
+          double
+          advect_forward_euler(std::vector<double> &volume_fractions,
+                               std::vector<Tensor<2,3> > &a_cosine_matrices,
+                               const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
```

Comment:
We usually place input parameters (`const` ones) before output parameters. Please reorder.